src/optimizer/cost_and_stats_calculator.cpp

```diff
@@ -10,127 +10,306 @@
+			for(size_t i = 0; i < sort_prop->GetSortColumnSize(); i++) {
+				oid_t column = (oid_t)((expression::TupleValueExpression *)sort_prop->GetSortColumn(i).get())->GetColumnId();
+				columns.push_back(column);
+				orders.push_back(sort_prop->GetSortAscending(i));
+			}
+			output_cost_ = Cost::OrderByCost(table_stats_ptr, columns, orders, output_stats);
+			output_stats_.reset(output_stats.get());
+		}
+		void CostAndStatsCalculator::Visit(const PhysicalLimit *) {
+			// TODO: Replace with more accurate cost
+			// lack of limit number
+			auto table_stats_ptr = std::dynamic_pointer_cast<TableStats>(child_stats_.at(0));
+			auto output_stats = generateOutputStat(table_stats_ptr, output_properties_
+				->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>());
+			size_t limit = (size_t) std::dynamic_pointer_cast<PropertyLimit>(
+				output_properties_->GetPropertyOfType(PropertyType::LIMIT))->GetLimit();
+			output_cost_ = Cost::LimitCost(std::dynamic_pointer_cast<TableStats>(child_stats_.at(0)),limit, output_stats);
+			output_stats_.reset(output_stats.get());
+		}
+		void CostAndStatsCalculator::Visit(const PhysicalFilter *){};
+		void CostAndStatsCalculator::Visit(const PhysicalInnerNLJoin *){
+			// TODO: Replace with more accurate cost
+			output_cost_ = 1;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalLeftNLJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalRightNLJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalOuterNLJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalInnerHashJoin *){
+			// TODO: Replace with more accurate cost
+			output_cost_ = 0;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalLeftHashJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalRightHashJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalOuterHashJoin *){};
+		void CostAndStatsCalculator::Visit(const PhysicalInsert *) {
+			// TODO: Replace with more accurate cost
+			output_cost_ = 0;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalInsertSelect *) {
+			// TODO: Replace with more accurate cost
+			output_cost_ = 0;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalDelete *) {
+			// TODO: Replace with more accurate cost
+			output_cost_ = 0;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalUpdate *) {
+			// TODO: Replace with more accurate cost
+			output_cost_ = 0;
+		};
+		void CostAndStatsCalculator::Visit(const PhysicalHashGroupBy * op) {
+			// TODO: Replace with more accurate cost
+			PL_ASSERT(child_stats_.size() == 1);
+			auto table_stats_ptr = std::dynamic_pointer_cast<TableStats>(child_stats_.at(0));
+			auto output_stats = generateOutputStat(table_stats_ptr, output_properties_
+				->GetPropertyOfType(PropertyType::COLUMNS)->As<PropertyColumns>());
+			std::vector<oid_t> column_ids;
+			for (auto column: op->columns) {
+				oid_t column_id = (oid_t)((expression::TupleValueExpression *)column.get())->GetColumnId();
```

Comment:
Why using .get() and type cast into raw pointer here but use std::dynamic_pointer_cast above?